File: health_advanced_swift/HealthDashboard/Views/HDSleepBarView.m
//
//  HDSleepBarView.m
//  HealthDashboard - 睡眠柱状图
//
//  Created by zhang, haizi on 2026/3/19.
//

#import "HDSleepBarView.h"

@implementation HDSleepBarView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        self.backgroundColor = [UIColor clearColor];
    }
    return self;
}

- (void)reloadData {
    // 移除旧子视图
    for (UIView *v in self.subviews) [v removeFromSuperview];

    if (!_hoursData.count) return;
    NSInteger count = MIN(7, (NSInteger)_hoursData.count);
    CGFloat w = self.bounds.size.width;
    CGFloat h = self.bounds.size.height;
    CGFloat barW = (w - 8 * (count - 1)) / count;
    CGFloat maxH = [_hoursData valueForKeyPath:@"@max.floatValue"] ? [[_hoursData valueForKeyPath:@"@max.floatValue"] floatValue] : 10.0;
    if (maxH < 1) maxH = 10;

    NSArray *days = @[@"一",@"二",@"三",@"四",@"五",@"六",@"日"];
    NSInteger startIdx = _hoursData.count - count;

    for (NSInteger i = 0; i < count; i++) {
        CGFloat hours = [_hoursData[startIdx + i] floatValue];
        CGFloat ratio = hours / maxH;
        CGFloat barH  = MAX(4, ratio * (h - 20));
        CGFloat x     = i * (barW + 8);

        // 柱子
        UIView *bar = [[UIView alloc] initWithFrame:CGRectMake(x, h - barH - 16, barW, barH)];
        bar.backgroundColor  = [UIColor colorWithRed:0.38 green:0.60 blue:0.95 alpha:1.0];
        bar.layer.cornerRadius = barW / 2;
        [self addSubview:bar];

        // 星期标签
        UILabel *lbl = [[UILabel alloc] initWithFrame:CGRectMake(x, h - 14, barW, 14)];
        lbl.text      = days[i % 7];
        lbl.font      = [UIFont systemFontOfSize:10];
        lbl.textColor = [UIColor colorWithWhite:0.55 alpha:1.0];
        lbl.textAlignment = NSTextAlignmentCenter;
        [self addSubview:lbl];
    }
}

- (void)layoutSubviews {
    [super layoutSubviews];
    [self reloadData];
}

@end

File: health_advanced_swift/HealthDashboard/Views/HDSleepBarView.swift
//
//  HDSleepBarView.swift
//  HealthDashboard - 睡眠柱状图
//
//  Created by AI Assistant on 2026/3/22.
//

import UIKit

/// 近7天睡眠横向柱状图
class HDSleepBarView: UIView {

    // MARK: - Properties

    var hoursData: [NSNumber] = []

    // MARK: - Init

    override init(frame: CGRect) {
        super.init(frame: frame)
        backgroundColor = .clear
    }

    required init?(coder: NSCoder) {
        super.init(coder: coder)
    }

    override func layoutSubviews() {
        super.layoutSubviews()
        reloadData()
    }

    // MARK: - Public

    func reloadData() {
        subviews.forEach { $0.removeFromSuperview() }
        guard !hoursData.isEmpty else { return }

        let count = min(7, hoursData.count)
        let w = bounds.width
        let h = bounds.height
        let barW = (w - 8 * CGFloat(count - 1)) / CGFloat(count)
        let maxH = hoursData.map { $0.floatValue }.max().map { CGFloat($0) } ?? 10
        let safeMaxH = maxH < 1 ? 10 : maxH
        let days = ["一", "二", "三", "四", "五", "六", "日"]
        let startIdx = hoursData.count - count

        for i in 0..<count {
            let hours = CGFloat(hoursData[startIdx + i].floatValue)
            let ratio = hours / safeMaxH
            let barH = max(4, ratio * (h - 20))
            let x = CGFloat(i) * (barW + 8)

            // 柱子
            let bar = UIView(frame: CGRect(x: x, y: h - barH - 16, width: barW, height: barH))
            bar.backgroundColor = UIColor(red: 0.38, green: 0.60, blue: 0.95, alpha: 1)
            bar.layer.cornerRadius = barW / 2
            addSubview(bar)

            // 星期标签
            let lbl = UILabel(frame: CGRect(x: x, y: h - 14, width: barW, height: 14))
            lbl.text = days[i % 7]
            lbl.font = .systemFont(ofSize: 10)
            lbl.textColor = UIColor(white: 0.55, alpha: 1)
            lbl.textAlignment = .center
            addSubview(lbl)
        }
    }
}
